File: packages/revcrm_accounts/src/models/Account.ts

import * as rev from 'rev-models';
import { EntityModel } from 'revcrm/lib/models/BaseModels';

export const ACCOUNT_TYPES = [
    ['organisation', 'Organisation'],
    ['contact', 'Contact']
];

export const TITLES = [
    ['Mr', 'Mr'],
    ['Mrs', 'Mrs'],
    ['Miss', 'Miss'],
    ['Ms', 'Ms'],
    ['Dr', 'Dr'],
    ['Sir', 'Sir'],
    ['Prof', 'Prof'],
];

export class Account extends EntityModel<Account> {

    @rev.SelectField({ label: 'Type', selection: ACCOUNT_TYPES })
        type: string;
    @rev.TextField({ label: 'Account Code' })
        code: string;
    @rev.TextField({ label: 'Organisation Name', required: false })
        org_name: string;
    @rev.SelectField({ label: 'Title', selection: TITLES, required: false })
        title: string;
    @rev.TextField({ label: 'First Name', required: false })
        first_name: string;
    @rev.TextField({ label: 'Last Name', required: false })
        last_name: string;

    @rev.TextField({ label: 'Name' })
        get name() {
            return this.type == 'organisation'
                ? this.org_name
                : (this.first_name || '') + ' ' + this.last_name;
        }

    @rev.TextField({ label: 'Phone', required: false })
        phone: string;
    @rev.TextField({ label: 'Mobile', required: false })
        mobile: string;
    @rev.TextField({ label: 'Fax', required: false })
        fax: string;
    @rev.TextField({ label: 'E-mail', required: false })
        email: string;
    @rev.URLField({ label: 'Website', required: false })
        website: string;
    @rev.TextField({ label: 'Notes', multiLine: true, required: false })
        notes: string;

    defaults() {
        this.type = 'organisation';
    }

    validate(vc: rev.IValidationContext) {
        if (this.type == 'organiation' && !this.org_name) {
            vc.result.addFieldError('org_name', 'Organisation Name is required');
        }
        if (this.type == 'contact' && !this.last_name) {
            vc.result.addFieldError('last_name', 'Last Name is required');
        }
    }

}

File: revcrm-server/src/main/kotlin/org/revcrm/entities/accounts/Account.kt
package org.revcrm.entities.accounts

import org.revcrm.annotations.Label
import org.revcrm.annotations.MultiLine
import org.revcrm.annotations.Validate
import org.revcrm.annotations.SelectionList
import org.revcrm.db.EntityValidationData
import org.revcrm.entities.Base
import xyz.morphia.annotations.Embedded
import xyz.morphia.annotations.Entity
import xyz.morphia.annotations.Index
import xyz.morphia.annotations.Indexes
import xyz.morphia.annotations.Field as IndexField
import xyz.morphia.utils.IndexType
import javax.validation.Valid

@Entity
@Indexes(
    Index(fields = [
        IndexField(value = "$**", type = IndexType.TEXT)
    ])
)
class Account(
    @Label("Is Company?")
    var is_company: Boolean,

    @Label("Company Name")
    var company_name: String?,

    @Label("Title")
    @SelectionList("contact_titles")
    var title: String?,

    @Label("First Name")
    var first_name: String?,

    @Label("Last Name")
    var last_name: String?,

    @Label("Account Code")
    var code: String?,

    @Label("Tags")
    @SelectionList("account_tags")
    var tags: String?,

    @Label("Source")
    @SelectionList("account_sources")
    var source: String?,

    @Label("Phone")
    var phone: String?,

    @Label("Mobile")
    var mobile: String?,

    @Label("Fax")
    var fax: String?,

    @Label("Email")
    var email: String?,

    @Label("Website")
    var website: String?,

    @Label("Notes")
    @MultiLine
    var notes: String?,

    @Label("Primary Address")
    @Embedded @field:Valid
    var primary_address: Address

) : Base() {

    @Label("Name")
    override val record_name: String
    get() {
        return if (is_company)
            company_name.toString()
        else
            "${title ?: ""} ${first_name ?: ""} ${last_name ?: ""}".trim()
    }

    @Validate
    fun validate(validation: EntityValidationData) {
        if (is_company && company_name.isNullOrBlank()) {
            validation.addFieldError(this,
                fieldPath = "company_name",
                errorCode = "NotBlank",
                message = "Company Name cannot be blank")
        } else if (!is_company && last_name.isNullOrBlank()) {
            validation.addFieldError(this,
                fieldPath = "last_name",
                errorCode = "NotBlank",
                message = "Contact Last Name cannot be blank")
        }
    }
}
